File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-option-shared-5-4/FxFeature.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class FxFeature : ISerialized
    {
        public FxFeature(XmlNode xmlNode)
        {
            XmlNode referenceCurrencyNode = xmlNode.SelectSingleNode("referenceCurrency");
            
            if (referenceCurrencyNode != null)
            {
                if (referenceCurrencyNode.Attributes["href"] != null || referenceCurrencyNode.Attributes["id"] != null) 
                {
                    if (referenceCurrencyNode.Attributes["id"] != null) 
                    {
                        referenceCurrencyIDRef_ = referenceCurrencyNode.Attributes["id"].Value;
                        IdentifiedCurrency ob = new IdentifiedCurrency(referenceCurrencyNode);
                        IDManager.SetID(referenceCurrencyIDRef_, ob);
                    }
                    else if (referenceCurrencyNode.Attributes["href"] != null)
                    {
                        referenceCurrencyIDRef_ = referenceCurrencyNode.Attributes["href"].Value;
                    }
                    else
                    {
                        referenceCurrency_ = new IdentifiedCurrency(referenceCurrencyNode);
                    }
                }
                else
                {
                    referenceCurrency_ = new IdentifiedCurrency(referenceCurrencyNode);
                }
            }
            
        
            XmlNode compositeNode = xmlNode.SelectSingleNode("composite");
            
            if (compositeNode != null)
            {
                if (compositeNode.Attributes["href"] != null || compositeNode.Attributes["id"] != null) 
                {
                    if (compositeNode.Attributes["id"] != null) 
                    {
                        compositeIDRef_ = compositeNode.Attributes["id"].Value;
                        Composite ob = new Composite(compositeNode);
                        IDManager.SetID(compositeIDRef_, ob);
                    }
                    else if (compositeNode.Attributes["href"] != null)
                    {
                        compositeIDRef_ = compositeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        composite_ = new Composite(compositeNode);
                    }
                }
                else
                {
                    composite_ = new Composite(compositeNode);
                }
            }
            
        
            XmlNode quantoNode = xmlNode.SelectSingleNode("quanto");
            
            if (quantoNode != null)
            {
                if (quantoNode.Attributes["href"] != null || quantoNode.Attributes["id"] != null) 
                {
                    if (quantoNode.Attributes["id"] != null) 
                    {
                        quantoIDRef_ = quantoNode.Attributes["id"].Value;
                        Quanto ob = new Quanto(quantoNode);
                        IDManager.SetID(quantoIDRef_, ob);
                    }
                    else if (quantoNode.Attributes["href"] != null)
                    {
                        quantoIDRef_ = quantoNode.Attributes["href"].Value;
                    }
                    else
                    {
                        quanto_ = new Quanto(quantoNode);
                    }
                }
                else
                {
                    quanto_ = new Quanto(quantoNode);
                }
            }
            
        
            XmlNode crossCurrencyNode = xmlNode.SelectSingleNode("crossCurrency");
            
            if (crossCurrencyNode != null)
            {
                if (crossCurrencyNode.Attributes["href"] != null || crossCurrencyNode.Attributes["id"] != null) 
                {
                    if (crossCurrencyNode.Attributes["id"] != null) 
                    {
                        crossCurrencyIDRef_ = crossCurrencyNode.Attributes["id"].Value;
                        Composite ob = new Composite(crossCurrencyNode);
                        IDManager.SetID(crossCurrencyIDRef_, ob);
                    }
                    else if (crossCurrencyNode.Attributes["href"] != null)
                    {
                        crossCurrencyIDRef_ = crossCurrencyNode.Attributes["href"].Value;
                    }
                    else
                    {
                        crossCurrency_ = new Composite(crossCurrencyNode);
                    }
                }
                else
                {
                    crossCurrency_ = new Composite(crossCurrencyNode);
                }
            }
            
        
        }
        
    
        #region ReferenceCurrency_
        private IdentifiedCurrency referenceCurrency_;
        public IdentifiedCurrency ReferenceCurrency_
        {
            get
            {
                if (this.referenceCurrency_ != null)
                {
                    return this.referenceCurrency_; 
                }
                else if (this.referenceCurrencyIDRef_ != null)
                {
                    referenceCurrency_ = IDManager.getID(referenceCurrencyIDRef_) as IdentifiedCurrency;
                    return this.referenceCurrency_; 
                }
                else
                {
                      return this.referenceCurrency_; 
                }
            }
            set
            {
                if (this.referenceCurrency_ != value)
                {
                    this.referenceCurrency_ = value;
                }
            }
        }
        #endregion
        
        public string referenceCurrencyIDRef_ { get; set; }
        #region Composite_
        private Composite composite_;
        public Composite Composite_
        {
            get
            {
                if (this.composite_ != null)
                {
                    return this.composite_; 
                }
                else if (this.compositeIDRef_ != null)
                {
                    composite_ = IDManager.getID(compositeIDRef_) as Composite;
                    return this.composite_; 
                }
                else
                {
                      return this.composite_; 
                }
            }
            set
            {
                if (this.composite_ != value)
                {
                    this.composite_ = value;
                }
            }
        }
        #endregion
        
        public string compositeIDRef_ { get; set; }
        #region Quanto_
        private Quanto quanto_;
        public Quanto Quanto_
        {
            get
            {
                if (this.quanto_ != null)
                {
                    return this.quanto_; 
                }
                else if (this.quantoIDRef_ != null)
                {
                    quanto_ = IDManager.getID(quantoIDRef_) as Quanto;
                    return this.quanto_; 
                }
                else
                {
                      return this.quanto_; 
                }
            }
            set
            {
                if (this.quanto_ != value)
                {
                    this.quanto_ = value;
                }
            }
        }
        #endregion
        
        public string quantoIDRef_ { get; set; }
        #region CrossCurrency_
        private Composite crossCurrency_;
        public Composite CrossCurrency_
        {
            get
            {
                if (this.crossCurrency_ != null)
                {
                    return this.crossCurrency_; 
                }
                else if (this.crossCurrencyIDRef_ != null)
                {
                    crossCurrency_ = IDManager.getID(crossCurrencyIDRef_) as Composite;
                    return this.crossCurrency_; 
                }
                else
                {
                      return this.crossCurrency_; 
                }
            }
            set
            {
                if (this.crossCurrency_ != value)
                {
                    this.crossCurrency_ = value;
                }
            }
        }
        #endregion
        
        public string crossCurrencyIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-option-shared-5-4/FxFeature.cpp
// FxFeature.cpp 
#include "FxFeature.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

FxFeature::FxFeature(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //referenceCurrencyNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* referenceCurrencyNode = xmlNode->FirstChildElement("referenceCurrency");

   if(referenceCurrencyNode){referenceCurrencyIsNull_ = false;}
   else{referenceCurrencyIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- referenceCurrencyNode , address : " << referenceCurrencyNode << std::endl;
   #endif
   if(referenceCurrencyNode)
   {
      if(referenceCurrencyNode->Attribute("href") || referenceCurrencyNode->Attribute("id"))
      {
          if(referenceCurrencyNode->Attribute("id"))
          {
             referenceCurrencyIDRef_ = referenceCurrencyNode->Attribute("id");
             referenceCurrency_ = boost::shared_ptr<IdentifiedCurrency>(new IdentifiedCurrency(referenceCurrencyNode));
             referenceCurrency_->setID(referenceCurrencyIDRef_);
             IDManager::instance().setID(referenceCurrencyIDRef_,referenceCurrency_);
          }
          else if(referenceCurrencyNode->Attribute("href")) { referenceCurrencyIDRef_ = referenceCurrencyNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { referenceCurrency_ = boost::shared_ptr<IdentifiedCurrency>(new IdentifiedCurrency(referenceCurrencyNode));}
   }

   //compositeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* compositeNode = xmlNode->FirstChildElement("composite");

   if(compositeNode){compositeIsNull_ = false;}
   else{compositeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- compositeNode , address : " << compositeNode << std::endl;
   #endif
   if(compositeNode)
   {
      if(compositeNode->Attribute("href") || compositeNode->Attribute("id"))
      {
          if(compositeNode->Attribute("id"))
          {
             compositeIDRef_ = compositeNode->Attribute("id");
             composite_ = boost::shared_ptr<Composite>(new Composite(compositeNode));
             composite_->setID(compositeIDRef_);
             IDManager::instance().setID(compositeIDRef_,composite_);
          }
          else if(compositeNode->Attribute("href")) { compositeIDRef_ = compositeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { composite_ = boost::shared_ptr<Composite>(new Composite(compositeNode));}
   }

   //quantoNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* quantoNode = xmlNode->FirstChildElement("quanto");

   if(quantoNode){quantoIsNull_ = false;}
   else{quantoIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- quantoNode , address : " << quantoNode << std::endl;
   #endif
   if(quantoNode)
   {
      if(quantoNode->Attribute("href") || quantoNode->Attribute("id"))
      {
          if(quantoNode->Attribute("id"))
          {
             quantoIDRef_ = quantoNode->Attribute("id");
             quanto_ = boost::shared_ptr<Quanto>(new Quanto(quantoNode));
             quanto_->setID(quantoIDRef_);
             IDManager::instance().setID(quantoIDRef_,quanto_);
          }
          else if(quantoNode->Attribute("href")) { quantoIDRef_ = quantoNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { quanto_ = boost::shared_ptr<Quanto>(new Quanto(quantoNode));}
   }

   //crossCurrencyNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* crossCurrencyNode = xmlNode->FirstChildElement("crossCurrency");

   if(crossCurrencyNode){crossCurrencyIsNull_ = false;}
   else{crossCurrencyIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- crossCurrencyNode , address : " << crossCurrencyNode << std::endl;
   #endif
   if(crossCurrencyNode)
   {
      if(crossCurrencyNode->Attribute("href") || crossCurrencyNode->Attribute("id"))
      {
          if(crossCurrencyNode->Attribute("id"))
          {
             crossCurrencyIDRef_ = crossCurrencyNode->Attribute("id");
             crossCurrency_ = boost::shared_ptr<Composite>(new Composite(crossCurrencyNode));
             crossCurrency_->setID(crossCurrencyIDRef_);
             IDManager::instance().setID(crossCurrencyIDRef_,crossCurrency_);
          }
          else if(crossCurrencyNode->Attribute("href")) { crossCurrencyIDRef_ = crossCurrencyNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { crossCurrency_ = boost::shared_ptr<Composite>(new Composite(crossCurrencyNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<IdentifiedCurrency> FxFeature::getReferenceCurrency()
{
   if(!this->referenceCurrencyIsNull_){
        if(referenceCurrencyIDRef_ != ""){
             return boost::shared_static_cast<IdentifiedCurrency>(IDManager::instance().getID(referenceCurrencyIDRef_));
        }else{
             return this->referenceCurrency_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<IdentifiedCurrency>();
   }
}
boost::shared_ptr<Composite> FxFeature::getComposite()
{
   if(!this->compositeIsNull_){
        if(compositeIDRef_ != ""){
             return boost::shared_static_cast<Composite>(IDManager::instance().getID(compositeIDRef_));
        }else{
             return this->composite_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Composite>();
   }
}
boost::shared_ptr<Quanto> FxFeature::getQuanto()
{
   if(!this->quantoIsNull_){
        if(quantoIDRef_ != ""){
             return boost::shared_static_cast<Quanto>(IDManager::instance().getID(quantoIDRef_));
        }else{
             return this->quanto_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Quanto>();
   }
}
boost::shared_ptr<Composite> FxFeature::getCrossCurrency()
{
   if(!this->crossCurrencyIsNull_){
        if(crossCurrencyIDRef_ != ""){
             return boost::shared_static_cast<Composite>(IDManager::instance().getID(crossCurrencyIDRef_));
        }else{
             return this->crossCurrency_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Composite>();
   }
}
}

